client/js/ifml/fromjson.js

```diff
@@ -6,197 +6,314 @@
-    switch (element.type) {
-    case 'ifml.ViewContainer':
-        return applyViewContainerMetadata(element, cells);
-    case 'ifml.ViewComponent':
-        return applyViewComponentMetadata(element, cells);
-    case 'ifml.Event':
-        return applyEventMetadata(element, cells);
-    case 'ifml.Action':
-        return applyActionMetadata(element, cells);
-    case 'ifml.NavigationFlow':
-        return applyNavigationFlowMetadata(element, cells);
-    case 'ifml.DataFlow':
-        return applyDataFlowMetadata(element, cells);
-    }
-}
-
-
-function applyRelation(relation, cells) {
-    switch (relation.type) {
-    case 'hierarchy':
-        cells[relation.parent].embed(cells[relation.child]);
-        break;
-    case 'source':
-        cells[relation.flow].set('source', {id: relation.source});
-        cells[relation.source].embed(cells[relation.flow]);
-        break;
-    case 'target':
-        cells[relation.flow].set('target', {id: relation.target});
-        break;
-    }
-}
+var jointElementRules = [
+    createRule(
+        function (element) {
+            return element.type === 'ifml.ViewContainer';
+        },
+        function (element) {
+            return new ifml.elements.ViewContainer(element);
+        }
+    ),
+    createRule(
+        function (element) {
+            return element.type === 'ifml.ViewComponent';
+        },
+        function (element) {
+            return new ifml.elements.ViewComponent(element);
+        }
+    ),
+    createRule(
+        function (element) {
+            return element.type === 'ifml.Event';
+        },
+        function (element) {
+            return new ifml.elements.Event(element);
+        }
+    ),
+    createRule(
+        function (element) {
+            return element.type === 'ifml.Action';
+        },
+        function (element) {
+            return new ifml.elements.Action(element);
+        }
+    ),
+    createRule(
+        function (element) {
+            return element.type === 'ifml.DataFlow';
+        },
+        function (element) {
+            return new ifml.links.DataFlow(element);
+        }
+    ),
+    createRule(
+        function (element) {
+            return element.type === 'ifml.NavigationFlow';
+        },
+        function (element) {
+            return new ifml.links.NavigationFlow(element);
+        }
+    )
+];
 
 exports.fromJSON = function (model) {
-    var cells = _.chain(model.elements)
-        .map(function (e) { return [e.id, mapElement(e)]; })
-        .filter(_.last)
-        .zipObject()
-        .value();
-
-    _.each(model.relations, function (r) {
-        applyRelation(r, cells);
-    });
 
-    _.each(model.elements, function (e) {
-        applyMetadata(e, cells);
-    });
+    var trasformer = createTransformer({
+            element: elementRules,
+            relation: relationRules
+        }, core.reduceBy('id', core.merge(core.last(), {embeds: core.concat()}))),
+        cells = createTransformer({
```

Comment:
The file is getting big, make `fromJSON` a folder and split the source in auxiliary files, one for each transformation and an index which does just orchestration.